cpu/lpc2387/periph/spi.c

```diff
@@ -0,0 +1,226 @@
+{
+    if (dev) {
+        return -1;
+    }
+
+    uint32_t   f_baud = 0;
+    switch(speed)
+    {
+    case SPI_SPEED_100KHZ:
+        f_baud = 100;
+        break;
+    case SPI_SPEED_400KHZ:
+        f_baud = 400;
+        break;
+    case SPI_SPEED_1MHZ:
+        f_baud = 1000;
+        break;
+    case SPI_SPEED_5MHZ:
+        f_baud = 5000;
+        break;
+    case SPI_SPEED_10MHZ:
+        f_baud = 10000;
+        break;
+    }
+
+#if 0
+    switch(conf)
+    {
+    case SPI_CONF_FIRST_RISING: /**< first data bit is transacted on the first rising SCK edge */
+        cpha = 0;
+        cpol = 0;
+        break;
+    case SPI_CONF_SECOND_RISING:/**< first data bit is transacted on the second rising SCK edge */
+        cpha = 1;
+        cpol = 0;
+        break;
+    case SPI_CONF_FIRST_FALLING:/**< first data bit is transacted on the first falling SCK edge */
+        cpha = 0;
+        cpol = 1;
+        break;
+    case SPI_CONF_SECOND_FALLING:/**< first data bit is transacted on the second falling SCK edge */
+        cpha = 1;
+        cpol = 1;
+        break;
+    }
+#endif
+
+    // Power
+    PCONP |= PCSSP0;                                // Enable power for SSP0 (default is on)
+
+    // PIN Setup
+    PINSEL3 |= BIT8 + BIT9;                         // Set CLK function to SPI
+    PINSEL3 |= BIT14 + BIT15;                       // Set MISO function to SPI
+    PINSEL3 |= BIT16 + BIT17;                       // Set MOSI function to SPI
+
+    // Interface Setup
+    SSP0CR0 = 7;
+
+    // Clock Setup
+    uint32_t pclksel;
+    uint32_t cpsr;
+    lpc2387_pclk_scale(F_CPU / 1000, f_baud, &pclksel, &cpsr);
+    PCLKSEL1 &= ~(BIT10 | BIT11);                       // CCLK to PCLK divider
+    PCLKSEL1 |= pclksel << 10;
+    SSP0CPSR = cpsr;
+
+    // Enable
+    SSP0CR1 |= BIT1;                                // SSP-Enable
+    int dummy;
+
+    // Clear RxFIFO:
+    while (SPI_RX_AVAIL) {                      // while RNE (Receive FIFO Not Empty)...
+        dummy = SSP0DR;                         // read data
+    }
+
+    /* to suppress unused-but-set-variable */
+    (void) dummy;
+    return 0;
+}
+
+int spi_init_slave(spi_t dev, spi_conf_t conf, char (*cb)(char))
+{
+    /* TODO */
+    return 0;
+}
+
+void spi_transmission_begin(spi_t dev, char reset_val)
+{
+    /* TODO*/
+}
+
+int spi_acquire(spi_t dev)
+{
+    if (dev >= SPI_NUMOF) {
+        return -1;
+    }
+    mutex_lock(&locks[dev]);
+    return 0;
+}
+
+int spi_release(spi_t dev)
+{
+    if (dev >= SPI_NUMOF) {
+        return -1;
+    }
+    mutex_unlock(&locks[dev]);
+    return 0;
+}
+
+int spi_transfer_byte(spi_t dev, char out, char *in)
+{
+    while (!SPI_TX_EMPTY);
+    SSP0DR = out;
+    while (SPI_BUSY);
+    while (!SPI_RX_AVAIL);
+
+    char tmp = (char)SSP0DR;
+
+    if (in != NULL) {
+        *in = tmp;
+    }
+
+    return 1;
+}
+
+int spi_transfer_bytes(spi_t dev, char *out, char *in, unsigned int length)
```

Comment:
this implementation is broken: when trying to use it full-duplex, it loses data...